src/dmclock_server.h

```diff
@@ -1762,13 +1764,14 @@ namespace crimson {
 
 	while (!this->finishing) {
 	  // predicate for cond.wait()
-	  const auto pred = [this] () -> bool { return this->finishing; };
+	  const auto pred = [this] () -> bool { return this->finishing || this->notified; };
```

Comment:
Why isn't this predicate simply this->finishing || sched_ahead_when != TimeZero?  Seems like adding notified is unnecessary.

Reply:
Also, I think you should combine the two PRs, reviewing these in isolation doesn't make a ton of sense as one modifies the cond variable predicate and the other modifies the loop behavior.

Reply:
@athanatos I think this->notified can be replaced with sched_ahead_when > TimeZero. I just confirmed  request_count() internally tries to hold the data_mtx lock, leading to dead lock issue. Is it okay?

I will merge https://github.com/ceph/dmclock/pull/86 with this PR.

Reply:
@athanatos I have made changes and cherry-picked #86 here.